### PR TITLE
Exact line search

### PR DESCRIPTION
**Pull request by phase**
Hello! Here is the current work on the exact line search.

### CODE REVIEW

**Comment by tdhock** [comment]
can you please add some tests/examples, and add your name in DESCRIPTION?
```r
Authors@R: c(
    person("Toby", "Hocking",
     email="toby.hocking@r-project.org",
     role=c("aut", "cre"),
    person("Jadon", "Fowler",
     role="aut",
     comment="Contributed exact line search C++ code")
```

File: DESCRIPTION [hunk 1]
```diff
@@ -2,7 +2,13 @@ Package: aum
 Type: Package
 Title: Area Under Minimum of False Positives and Negatives
 Version: 2022.2.7
-Author: Toby Dylan Hocking
+Authors@R: c(
+    person("Toby", "Hocking",
+     email="toby.hocking@r-project.org",
+     role=c("aut", "cre"),
```
Comment (tdhock) [inline DESCRIPTION]: @phase  please make sure the parens are matched

File: src/interface.cpp [hunk 2]
```diff
@@ -59,3 +60,41 @@ Rcpp::List aum_sort_interface
...
+            &fpDiff[0],
+            &fnDiff[0],
+            initialAum,
+            maxIterations,
+            &FP[0], &FN[0], &M[0],
+            &stepSizeVec[0],
+            &aumVec[0]
+    );
+
+    return Rcpp::List::create(Rcpp::Named("aum", aumVec), Rcpp::Named("step.size", stepSizeVec));
```
Comment (tdhock) [inline src/interface.cpp]: please return Rcpp::DataFrame instead of List since these two vectors are the same size

**Comment by tdhock** [comment]
hi @phase I added a plot method which gives results like below,
image
I was thinking, it would be useful to see what happens after the last step size (with the current output we don't know what is the slope of aum past the largest step size). Is it possible to add another row to the output, with step.size=Inf and an aum value? 

* aum=some finite value (same as previous aum value) if there is no change, slope of zero, after the largest step size. (as in plot above)
* aum=Inf if the slope is positive after the largest step size.

**Comment by tdhock** [comment]
I added a plot method for new function aum_line_search_grid which can be used to check the calculation. 
For a simple example it shows that the algo works:
image

For the more complicated example you are working with it shows that the algo is not working. 
image
I only ran the algo for 30 iterations, and I used 100 grid points (blue) to check the calculation. It seems there are some problem with computing 

* the intersection points. Clearly there are quite a few intersections happening between 0 and 0.1, but the first intersection point is detected somewhere between 0.1 and 0.2 (first grey vertical line), so this suggests a bug in the C++ code.
* the initial slope. clearly the green points should line up on the first line segment after step size 0, even if you don't get the first intersection point right.

About the consistency between penaltyLearning and aum packages calculation of the AUM, aum package assumes FP=0 at -Inf and FN=0 at Inf, and only inputs the fp_diff and fn_diff. However penaltyLearning can handle the opposite as well (FP=0 at Inf etc). For changepoint problems using aum package we define predicted value = - log(lambda) whereas typically with penaltyLearning package we use pred.log.lambda (and do not take the negative).

**Comment by tdhock** [comment]
code below shows how you have to take the negative to get consistent results between penaltyLearning and aum packages,
```r
data(neuroblastomaProcessed, package="penaltyLearning", envir=environment())
library(data.table)
some.ids <- c("1.2", "1.1", "4.1", "4.2")
nb.err <- with(neuroblastomaProcessed$errors, data.table(
  example=paste0(profile.id, ".", chromosome),
  min.log.lambda,
  max.log.lambda,
  fp, fn))[example %in% some.ids]
nb.segs <- nb.err[, seg.i := cumsum(
  c(1, diff(fp)!=0 | diff(fn) != 0)),
  by=example
  ][, .(
  min.log.lambda=min(min.log.lambda),
  max.log.lambda=max(max.log.lambda),
  fp=fp[1],
  fn=fn[1]
  ), by=.(example, seg.i)]
(nb.diffs <- aum::aum_diffs_penalty(
  nb.err[,min.lambda := exp(min.log.lambda)], some.ids))
pred.dt <- data.table(
  example=some.ids,
  pred.log.lambda=seq(1, length(some.ids)))
roc.list <- penaltyLearning::ROChange(nb.segs[, `:=`(
  possible.fp = max(fp),
  possible.fn = max(fn),
  labels=1,
  errors=fp+fn
)], pred.dt, problem.vars="example")
roc.list$aum
aum::aum(nb.diffs, pred.dt$pred.log.lambda)$aum
aum::aum(nb.diffs, -pred.dt$pred.log.lambda)$aum
```
output:
```r
> roc.list$aum
[1] 0.5180666
> aum::aum(nb.diffs, pred.dt$pred.log.lambda)$aum
[1] 0
> aum::aum(nb.diffs, -pred.dt$pred.log.lambda)$aum
[1] 0.5180666
> nb.segs
    example seg.i min.log.lambda max.log.lambda    fp    fn
     <char> <num>          <num>          <num> <num> <num>
 1:     1.1     1      0.1957271            Inf     0     0
 2:     1.1     2           -Inf      0.1957271     1     0
 3:     1.2     1     -1.1087554            Inf     0     0
 4:     1.2     2           -Inf     -1.1087554     1     0
 5:     4.1     1      3.2651854            Inf     0     1
 6:     4.1     2           -Inf      3.2651854     0     0
 7:     4.2     1      1.9293005            Inf     0     1
 8:     4.2     2      1.3880728      1.9293005     0     0
 9:     4.2     3      1.1364327      1.3880728     0     1
10:     4.2     4           -Inf      1.1364327     0     0
> nb.diffs
   example       pred fp_diff fn_diff
     <int>      <num>   <num>   <num>
1:       1 -0.1957271       1       0
2:       0  1.1087554       1       0
3:       2 -3.2651854       0      -1
4:       3 -1.9293005       0      -1
5:       3 -1.3880728       0       1
6:       3 -1.1364327       0      -1
```

**Comment by tdhock** [comment]
the problem was that the input DF to the exact line search needs to be sorted by intercept. for the simple examples it was sorted, but for the more complicated one it was not. I added some R code which sorts by intercept before calling the C++ code, and the result looks good!
image

**Comment by phase** [comment]
hey @tdhock, thanks a ton for that!

I switched the training loop to use `aum::aum_line_search_grid()` and I'm still getting negative AUM values
(cyan dots here are from `aum.result$grid_aum`) 
image


[here](https://github.com/phase/max-generalized-auc/blob/56902ffc2510ec6e72688cdcd4b600d16ac37bb0/figure-aum-optimized-data.R#L107-L159) is my code for the loop. Still not sure why the points are going down. Does `diff.fp.fn` need to get sorted differently between iterations?

> For changepoint problems using aum package we define predicted value = - log(lambda) whereas typically with penaltyLearning package we use pred.log.lambda (and do not take the negative).

If I negate `pred.log.lambda`, then the AUM doesn't decrease unless I also negate the gradient here:
```diff
-  step.pred$pred.log.lambda <- step.pred$pred.log.lambda - (smallest.aum$step.size * aum.result$gradient)
-  step.aum.result <- aum::aum_line_search_grid(diff.fp.fn, pred.vec=step.pred$pred.log.lambda, maxIterations=max.iterations)
+  step.pred$pred.log.lambda <- step.pred$pred.log.lambda - (smallest.aum$step.size * -aum.result$gradient)
+  step.aum.result <- aum::aum_line_search_grid(diff.fp.fn, pred.vec=-step.pred$pred.log.lambda, maxIterations=max.iterations)
```

and this only does 2 iterations:
image

Going to continue investigating!

**Comment by tdhock** [comment]
I suspect there is some bug in the C++ code that only appears when the input data are large/complex.
I think you should try to find a simple example for which the bug is evident, and then insert print statements in your C++ code to see what is going on.
For example here is an example where the algo works in the beginning but then starts to show problems after iteration number 192.
```r
  data(neuroblastomaProcessed, package="penaltyLearning", envir=environment())
  nb.err <- with(neuroblastomaProcessed$errors, data.frame(
    example=paste0(profile.id, ".", chromosome),
    min.lambda,
    max.lambda,
    fp, fn))
  all.ids <- rownames(neuroblastomaProcessed$feature.mat[1:800,])
  all.diffs <- aum::aum_diffs_penalty(nb.err, all.ids)
  current.pred <- rep(0, length(all.ids))
  nb.all.search <- aum::aum_line_search_grid(all.diffs, pred.vec=current.pred, n.grid=100)
  if(requireNamespace("ggplot2"))plot(nb.all.search)
  compare.dt <- with(nb.all.search, grid_aum[
    data.table(line_search_result)[, iteration := .I],
    .(iteration, step.size, grid.aum=x.aum, exact.aum=i.aum, diff=x.aum-i.aum),
    on="step.size"])
```
I get:
```r
> compare.dt[190:200]
    iteration  step.size grid.aum exact.aum          diff
        <int>      <num>    <num>     <num>         <num>
 1:       190 0.09689024 28.95784  28.95784  0.000000e+00
 2:       191 0.09700172 28.95383  28.95383 -7.105427e-15
 3:       192 0.09850517 28.90259  28.89971  2.879866e-03
 4:       193 0.09940085 28.87213  28.86746  4.671224e-03
 5:       194 0.10067382 28.82885  28.82163  7.217162e-03
 6:       195 0.10089931 28.82119  28.81352  7.668129e-03
 7:       196 0.10220985 28.77663  28.76634  1.028922e-02
 8:       197 0.10269704 28.76006  28.74880  1.126360e-02
 9:       198 0.10280873 28.75627  28.74478  1.148698e-02
10:       199 0.10341980 28.73549  28.72278  1.270912e-02
11:       200 0.10372154 28.72523  28.71192  1.331259e-02
```
grid.aum should equal exact aum column, and diff should equal zero, but starts to become non-zero at iteration 192.

**Comment by tdhock** [comment]
looks like the bug happens when there are some input values with the exact same values of slope and intercept.
```r
  all.ids <- rownames(neuroblastomaProcessed$feature.mat[1:800,])
  all.diffs <- aum::aum_diffs_penalty(nb.err, all.ids)
  current.pred <- rep(0, length(all.ids))
  nb.all.search <- aum::aum_line_search_grid(all.diffs, pred.vec=current.pred, n.grid=100)
  all.slope.int <- data.table(nb.all.search$line_search_input)[, .(slope, intercept)]
  uniq.slope.int <- unique(all.slope.int)
  rbind(all=nrow(all.slope.int), uniq=nrow(uniq.slope.int))
```
output:
```r
> rbind(all=nrow(all.slope.int), uniq=nrow(uniq.slope.int))
     [,1]
all   806
uniq  800
```

**Comment by tdhock** [comment]
Although this could be fixed in the C++ code, here I propose a fix in the R code, by summing fp/fn diffs before passing them as input to the exact line search.

**Comment by phase** [comment]
I pulled your changes and the examples you have worked fine for me, but my original example is still showing errors:
image
```
step=1 size=5.318526e-01 aum=217.236143->133.884796
step=2 size=3.358806e-01 aum=133.884796->111.020869
step=3 size=2.271783e-01 aum=111.020869->97.761572
step=4 size=6.269758e-02 aum=97.761572->93.784970
step=5 size=3.198197e-02 aum=93.784970->92.201766
step=6 size=1.198343e+01 aum=92.201766->1450.207745
```
here's a plot of that last result:
image

File: src/aum_line_search.cpp [hunk 3]
```diff
@@ -43,7 +43,7 @@ void queueIntersection(
     );
 
     // intersection points with infinite values aren't real intersections
-    if (intersectionPoint.isFinite() && intersectionPoint.x >= currentStepSize) {
+    if (intersectionPoint.isFinite() && intersectionPoint.x > currentStepSize) {
```
Comment (tdhock) [inline src/aum_line_search.cpp:46]: Using > instead of >= here fixed a bug for me, does that help your case?

**Comment by phase** [comment]
> Using > instead of >= here fixed a bug for me, does that help your case?

that worked!!! 

<details><summary>aum steps</summary>
<p>
<pre><code>
step=1 size=5.318526e-01 aum=217.236143->133.884796
step=2 size=3.358806e-01 aum=133.884796->111.020869
step=3 size=2.271783e-01 aum=111.020869->97.761572
step=4 size=6.269758e-02 aum=97.761572->93.784970
step=5 size=3.198197e-02 aum=93.784970->92.201766
step=6 size=1.939819e-02 aum=92.201766->91.208364
step=7 size=1.150924e-02 aum=91.208364->90.665276
step=8 size=6.875106e-03 aum=90.665276->90.386881
step=9 size=5.330056e-03 aum=90.386881->90.170987
step=10 size=4.690934e-03 aum=90.170987->90.003676
step=11 size=3.737403e-03 aum=90.003676->89.867379
step=12 size=1.668327e-03 aum=89.867379->89.785067
step=13 size=1.340681e-03 aum=89.785067->89.728747
step=14 size=1.203223e-03 aum=89.728747->89.686472
step=15 size=7.927493e-04 aum=89.686472->89.654585
step=16 size=3.785559e-04 aum=89.654585->89.626042
step=17 size=1.548017e-04 aum=89.626042->89.623357
step=18 size=1.317720e-04 aum=89.623357->89.615158
step=19 size=1.021072e-04 aum=89.615158->89.611415
step=20 size=1.021072e-04 aum=89.611415->89.607653
step=21 size=1.528803e-04 aum=89.607653->89.604930
step=22 size=2.383501e-04 aum=89.604930->89.599564
step=23 size=1.315680e-04 aum=89.599564->89.594201
step=24 size=2.105089e-04 aum=89.594201->89.588006
step=25 size=1.420935e-04 aum=89.588006->89.585589
step=26 size=1.136748e-04 aum=89.585589->89.578196
step=27 size=7.578320e-05 aum=89.578196->89.574894
step=28 size=3.264347e-03 aum=89.574894->89.570608
step=29 size=4.380428e-03 aum=89.570608->89.415948
step=30 size=2.776707e-03 aum=89.415948->89.351300
step=31 size=2.018208e-03 aum=89.351300->89.232373
step=32 size=1.389322e-03 aum=89.232373->89.182918
step=33 size=6.177872e-04 aum=89.182918->89.154431
step=34 size=4.242141e-04 aum=89.154431->89.142451
step=35 size=3.393713e-04 aum=89.142451->89.125825
step=36 size=2.036776e-04 aum=89.125825->89.123643
step=37 size=2.447479e-04 aum=89.123643->89.112499
step=38 size=1.637974e-04 aum=89.112499->89.109108
step=39 size=1.436355e-04 aum=89.109108->89.103130
step=40 size=8.421371e-05 aum=89.103130->89.099107
step=41 size=7.325536e-05 aum=89.099107->89.096868
step=42 size=1.055643e-04 aum=89.096868->89.094092
step=43 size=4.983761e-05 aum=89.094092->89.092514
step=44 size=3.197300e-05 aum=89.092514->89.091314
step=45 size=2.485334e-05 aum=89.091314->89.090718
step=46 size=2.409234e-05 aum=89.090718->89.089650
step=47 size=8.828013e-06 aum=89.089650->89.089220
step=48 size=7.205345e-06 aum=89.089220->89.088900
step=49 size=5.152042e-06 aum=89.088900->89.088745
step=50 size=3.572657e-06 aum=89.088745->89.088641
step=51 size=2.342922e-06 aum=89.088641->89.088508
step=52 size=1.631255e-06 aum=89.088508->89.088454
step=53 size=1.070712e-06 aum=89.088454->89.088429
step=54 size=1.069935e-06 aum=89.088429->89.088384
step=55 size=4.794734e-07 aum=89.088384->89.088367
step=56 size=4.234958e-07 aum=89.088367->89.088344
step=57 size=1.468048e-07 aum=89.088344->89.088341
step=58 size=1.445611e-07 aum=89.088341->89.088334
step=59 size=9.703120e-08 aum=89.088334->89.088331
step=60 size=5.521643e-08 aum=89.088331->89.088330
</code></pre>
</p>
</details>

File: src/aum_line_search.cpp [hunk 4]
```diff
@@ -0,0 +1,180 @@
+#include "aum_line_search.h"
+
+using namespace std;
+
+bool Point::operator==(const Point other) const {
+    return fabs(x - other.x) < 1e-6 && fabs(y - other.y) < 1e-6;
```
Comment (tdhock) [inline src/aum_line_search.cpp:6]: would be better if this constant could be user-specified or chosen based on the data, but this seems to work.

**Comment by tdhock** [comment]
hi you are collaborator on tdhock/aum so can you please make a branch in that repo next time? (instead of from your fork?)

**Pull request merged**